src/Autofac/Util/InternalTypeExtensions.cs

```diff
@@ -214,6 +213,33 @@ public static bool IsOpenGenericTypeOf(this Type @this, Type type)
           || @this.CheckInterfacesAreOpenGenericTypeOf(type);
     }
 
+    /// <summary>
+    /// Filters off interfaces, abstract types, and generally non-registerable
+    /// types. Largely used during type/assembly scanning, but also can
+    /// determine if this is something we can activate by reflection.
+    /// Intentionally does NOT filter out open generic definitions because this
+    /// gets used in both concrete and open generic scanning.
+    /// </summary>
+    /// <param name="type">
+    /// The type to check.
+    /// </param>
+    /// <param name="allowCompilerGenerated">
+    /// <see langword="true"/> to allow compiler-generated types to be
+    /// considered registerable. Defaults to <see langword="false"/>.
+    /// </param>
+    /// <returns>
+    /// <see langword="true"/> if the type is allowed to be registered during
+    /// scanning based on its reflection attributes; otherwise <see
+    /// langword="false"/>.
+    /// </returns>
+    // Issue #897: For back compat reasons we can't filter out
+    // non-public types here. Folks use assembly scanning on their
+    // own stuff, so encapsulation is a tricky thing to manage.
+    // If people want only public types, a LINQ Where clause can be used.
+    //
+    // Run IsCompilerGenerated check last due to perf. See AssemblyScanningPerformanceTests.MeasurePerformance.
+    internal static bool MayAllowReflectionActivation(this Type? type, bool allowCompilerGenerated = false) => type != null && type.IsClass && !type.IsAbstract && !type.IsDelegate() && (allowCompilerGenerated || !type.IsCompilerGenerated());
```

Comment:
This gets used a lot - ACTNARS, `RegisterTypes`, `RegistrationBuilder.FromType`, assembly scanning, open generic assembly scanning. The permutations are challenging because sometimes (open generic scanning) you want open generics, but usually you don't; sometimes (ACTNARS) you want compiler generated, but usually you don't.

But at least this gives us a single spot for doing these checks and ensuring the proper order.